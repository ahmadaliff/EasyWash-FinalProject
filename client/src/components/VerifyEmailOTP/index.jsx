import PropTypes from 'prop-types';
import Countdown from 'react-countdown';
import { useForm } from 'react-hook-form';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { FormattedMessage, injectIntl } from 'react-intl';

import InputRHF from '@components/InputRHF';

import { selectTokenEmail } from '@containers/Client/selectors';
import { selectEmail, selectExpire, selectStep } from '@pages/Register/selectors';
import { actionSetStep } from '@pages/Register/actions';
import { actionHandleSendEmailVerify, actionHandleSendOTP } from '@containers/Client/actions';

import classes from '@components/VerifyEmailOTP/style.module.scss';

const VerifyEmailOTP = ({ tokenVerify, isExpire, email, step, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const {
    handleSubmit,
    register,
    formState: { errors },
  } = useForm();

  const onSubmit = (data) => {
    data.token = tokenVerify;
    dispatch(actionHandleSendOTP(data));
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)} className={classes.form} data-testid="verify-email-otp-form">
      <InputRHF
        input={{
          name: 'otp',
          required: formatMessage({ id: 'app_user_otp_require_message' }),
          type: 'number',
          label: formatMessage({ id: 'app_user_otp' }),
          minLength: 4,
          messageMin: formatMessage({ id: 'app_user_otp' }),
        }}
        register={register}
        errors={errors}
      />

      <div className={classes.timerComp}>
        <FormattedMessage id="app_resend" /> ?
        <Countdown date={isExpire} data-testid="countdown-timer">
          <button
            type="button"
            onClick={() => {
              dispatch(actionHandleSendEmailVerify({ email }));
              dispatch(actionSetStep(1));
            }}
            className={classes.buttonResend}
            data-testid="button-resend"
          >
            <FormattedMessage id="app_resend" />
          </button>
        </Countdown>
      </div>
      <div className={classes.buttonWrap}>
        <button
          type="button"
          className={classes.buttonSubmit}
          onClick={() => dispatch(actionSetStep(step - 1))}
          data-testid="button-back"
        >
          <FormattedMessage id="app_back" />
        </button>
        <button type="submit" className={classes.buttonSubmit} data-testid="button-submit">
          <FormattedMessage id="app_next" />
        </button>
      </div>
    </form>
  );
};

VerifyEmailOTP.propTypes = {
  intl: PropTypes.object,
  step: PropTypes.number,
  tokenVerify: PropTypes.string,
  email: PropTypes.string,
  isExpire: PropTypes.number,
};

const mapStateToProps = createStructuredSelector({
  tokenVerify: selectTokenEmail,
  step: selectStep,
  email: selectEmail,
  isExpire: selectExpire,
});

export default injectIntl(connect(mapStateToProps)(VerifyEmailOTP));
