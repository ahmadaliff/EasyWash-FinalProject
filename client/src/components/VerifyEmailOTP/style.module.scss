@import '@styles/base';

.form {
  .buttonWrap {
    @include directionX;
    gap: 1rem;
    .buttonSubmit {
      width: 100%;
      border-radius: 2.5rem;
      background-color: var(--color-primary);
      color: var(--color-bg);
      padding: 0.5rem 2.5rem 0.5rem 2.5rem;
      margin: auto;
      margin-top: 0.75rem;
      box-shadow: none;
      border: none;
      cursor: pointer;
      &:hover {
        transform: scale(1.0125, 1.0125);
      }
      &:disabled {
        background-color: rgba(gray, 0.2);
        color: rgba($color: gray, $alpha: 0.5);
      }
      &:disabled:hover {
        transform: none;
      }
    }
  }
  .timerComp {
    font-size: small;
    padding: 0.5rem 0 0.5rem 0;
    @include directionX;
    align-items: center;
    gap: 0.5rem;
  }
  .buttonResend {
    border: none;
    border-radius: 10rem;
    padding: 0.25rem 1rem 0.25rem 1rem;
    background-color: var(--color-primary);
    color: var(--color-bg);
    cursor: pointer;
    font-size: smaller;
    &:hover {
      transform: scale(1.05, 1.05);
    }
  }
}
.showPassword {
  @include directionX;
  align-items: center;
  padding-top: 0.25rem;
  font-size: small;
}
