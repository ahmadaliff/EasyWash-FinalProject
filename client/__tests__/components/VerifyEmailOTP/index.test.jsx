import { render, fireEvent, screen } from '@utils/testHelper';
import { act } from 'react-dom/test-utils';
import VerifyEmailOTP from '@components/VerifyEmailOTP';

let wrapper;
const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));

jest.mock('@pages/Register/selectors', () => ({
  selectEmail: () => 'test@example.com',
  selectExpire: () => Date.now() + 1000,
  selectStep: () => 2,
}));

jest.mock('@containers/Client/selectors', () => ({
  selectTokenEmail: () => 'token',
}));

const mockProps = {
  intl: {
    formatMessage: jest.fn(),
  },
  tokenVerify: 'mockToken',
};

beforeEach(() => {
  act(() => {
    wrapper = render(<VerifyEmailOTP {...mockProps} />);
  });
  jest.clearAllMocks();
});

describe('VerifyEmailOTP Component', () => {
  test('Correct render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('verify-email-otp-form')).toBeInTheDocument();
    expect(screen.getByLabelText('OTP')).toBeInTheDocument();
    expect(getByTestId('button-back')).toBeInTheDocument();
    expect(getByTestId('button-submit')).toBeInTheDocument();
  });

  test('Handles form submission with correct data', async () => {
    const { getByTestId } = wrapper;
    const inputOTP = screen.getByLabelText('OTP');
    const submitButton = getByTestId('button-submit');

    fireEvent.change(inputOTP, { target: { value: '1234' } });

    await act(() => {
      fireEvent.click(submitButton);
    });
    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });
  test('Handles back', async () => {
    const { getByTestId } = wrapper;
    const backButton = getByTestId('button-back');
    await act(() => {
      fireEvent.click(backButton);
    });
    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });

  test('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
