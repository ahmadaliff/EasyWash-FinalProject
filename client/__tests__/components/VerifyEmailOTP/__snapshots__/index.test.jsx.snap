// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerifyEmailOTP Component Matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="form"
        data-testid="verify-email-otp-form"
      >
        <div
          class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
          data-testid="inputRHF"
        >
          <label
            class="inputLabel"
            for="otp"
          >
            OTP
            <input
              class="input undefined"
              id="otp"
              name="otp"
              type="number"
              value=""
            />
          </label>
        </div>
        <div
          class="timerComp"
        >
          Kirim Ulang
           ?
          <span>
            00
            :
            00
            :
            00
            :
            01
          </span>
        </div>
        <div
          class="buttonWrap"
        >
          <button
            class="buttonSubmit"
            data-testid="button-back"
            type="button"
          >
            Kembali
          </button>
          <button
            class="buttonSubmit"
            data-testid="button-submit"
            type="submit"
          >
            Lanjut
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="form"
      data-testid="verify-email-otp-form"
    >
      <div
        class="MuiFormControl-root inputContainer css-1nrlq1o-MuiFormControl-root"
        data-testid="inputRHF"
      >
        <label
          class="inputLabel"
          for="otp"
        >
          OTP
          <input
            class="input undefined"
            id="otp"
            name="otp"
            type="number"
            value=""
          />
        </label>
      </div>
      <div
        class="timerComp"
      >
        Kirim Ulang
         ?
        <span>
          00
          :
          00
          :
          00
          :
          01
        </span>
      </div>
      <div
        class="buttonWrap"
      >
        <button
          class="buttonSubmit"
          data-testid="button-back"
          type="button"
        >
          Kembali
        </button>
        <button
          class="buttonSubmit"
          data-testid="button-submit"
          type="submit"
        >
          Lanjut
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
